iroh-resolver/src/hamt.rs

```diff
@@ -68,21 +87,66 @@ impl InnerNode {
     ) -> Result<Self> {
         let path = Path::from_cid(link.cid);
         let out = loader.resolve(path).await?;
+
         match out.content {
-            OutContent::Unixfs(node) => match node {
-                UnixfsNode::HamtShard(ref shard) => {
-                    let node = Node::from_node(shard)?;
-                    Ok(InnerNode::Node(node))
-                }
-                UnixfsNode::File(_) => Ok(InnerNode::Leaf {
+            OutContent::Unixfs(value) => match value {
+                UnixfsNode::HamtShard(_, ref hamt) => Ok(InnerNode::Node {
+                    node: hamt.root.clone(),
+                    value,
+                }),
+                UnixfsNode::RawNode(_)
+                | UnixfsNode::File(_)
+                | UnixfsNode::Directory(_)
+                | UnixfsNode::Raw(_)
+                | UnixfsNode::Symlink(_) => Ok(InnerNode::Leaf {
                     link: link.clone(),
-                    value: node,
+                    value,
                 }),
-                _ => bail!("unexpected unixfs node: {:?}", node.typ()),
             },
+            OutContent::Raw(_, bytes) => {
+                let node = UnixfsNode::decode(&link.cid, bytes)?;
+
+                Ok(InnerNode::Leaf {
+                    link: link.clone(),
+                    value: node,
+                })
+            }
             _ => bail!("unexpected node: {:?}", out.content.typ()),
         }
     }
+    fn children<'a, 'b: 'a, C: ContentLoader>(
+        &'a self,
+        loader: &'b Resolver<C>,
+    ) -> impl Stream<Item = Result<Link>> + 'a {
+        async_stream::try_stream! {
+            match self {
+                InnerNode::Node { node, .. } => {
+                    let mut children = node.children(loader);
+                    while let Some(link) = children.next().await {
+                        let link = link?;
+                        yield link;
+                    }
+
+                },
+                InnerNode::Leaf { value, .. } => match value {
+                    UnixfsNode::Directory(_) => {
+                        for link in value.links() {
+                            let link = link?;
+                            yield link.to_owned();
+                        }
+                    }
+                    UnixfsNode::HamtShard(_, hamt) => {
+                        let mut children = hamt.children(loader);
+                        while let Some(link) = children.next().await {
+                            let link = link?;
+                            yield link;
+                        }
+                    }
+                    _ => {}
```

Comment:
why is it expected that only the `UnixfsNode::Directory` and `UnixfsNode::HamtShard` will have children? Isn't it possible for `UnixfsNode::File`s to also have links?

I'm probably missing a subtly here.

Reply:
The subtly here is my bad naming skillz. "children" here means things in a directory, not children as in links